src/rt/aaA.d

```diff
@@ -142,16 +142,36 @@ private:
     {
         auto obuckets = buckets;
         buckets = allocBuckets(ndim);
+        uint newFirstUsed = cast(uint)ndim;
 
-        foreach (ref b; obuckets)
+        foreach (ref b; obuckets[firstUsed..$])
             if (b.filled)
-                *findSlotInsert(b.hash) = b;
+            {
+                auto x = findSlotInsert(b.hash);
+                newFirstUsed = min(newFirstUsed, cast(uint)(x - buckets.ptr));
+                *x = b;
+            }
 
-        firstUsed = 0;
+        firstUsed = newFirstUsed;
         used -= deleted;
         deleted = 0;
         GC.free(obuckets.ptr); // safe to free b/c impossible to reference
     }
+
+    void clear() pure nothrow
+    {
+        foreach(ref b; buckets[firstUsed..$])
+        {
+            if(b.filled)
+            {
+                b.hash = HASH_DELETED;
+                b.entry = null;
+            }
+        }
+        deleted = used;
+        firstUsed = cast(uint)dim;
```

Comment:
Since when?  The phobos codebase is roughtly 50/50 on this, and the dmd codebase consistently uses no space after cast.